### PR TITLE
Use macro for button buttons

### PR DESCRIPTION
**Pull request by Katrienvh**
## Type
- Enhancement

## Pull request description
All buttons with text and icon are using the same macro now.

### CODE REVIEW

File: src/Backend/Modules/Blog/Layout/Templates/EditComment.html.twig [hunk 1]
```diff
@@ -4,7 +4,7 @@
 {% block actionbar %}
   <div class="btn-toolbar pull-right">
     <div class="btn-group" role="group">
-      {{ macro.buttonIcon( SITE_URL ~ itemURL, 'eye', 'lbl.View'|trans|ucfirst, '', "target=_blank") }}
+      {{ macro.buttonIcon( SITE_URL ~ itemURL, 'eye', 'lbl.View'|trans|ucfirst, '', { "target":"_blank" }) }}
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Blog/Layout/Templates/EditComment.html.twig]: I would keep it consistent and remove the spaces `( here )` and `{ here }`

File: src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig [hunk 2]
```diff
@@ -78,17 +78,11 @@
               <div class="btn-group pull-left" role="group">
                 {% if showMediaLibraryMediaFolderDelete %}
                   {% if not mediaFolder.hasConnectedItems and not mediaFolder.hasChildrenWithConnectedItems %}
-                    <button type="button" class="btn btn-danger" id="media-folder-delete" data-toggle="modal" data-target="#confirmMediaFolderDelete">
-                      {{ macro.icon('trash-o') }}
-                      {{ 'lbl.MediaFolderDelete'|trans|ucfirst }}
-                    </button>
+                    {{ macro.buttonIcon('', 'trash-o', 'lbl.MediaFolderDelete'|trans|ucfirst, 'btn-danger', { "id":"media-folder-delete", "data-toggle":"modal", "data-target":"#confirmMediaFolderDelete", "type":"button" } ) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig]: All id's are camelcased, so this one should be too.

File: src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig [hunk 3]
```diff
@@ -78,17 +78,11 @@
...
                   {% endif %}
                 {% endif %}
                 {% if showMediaLibraryMediaFolderEdit %}
-                  <button type="button" class="btn" id="media-folder-edit" data-toggle="modal" data-target="#editMediaFolderDialog">
-                    {{ macro.icon('pencil') }}
-                    {{ 'lbl.EditMediaFolder'|trans|ucfirst }}
-                  </button>
+                  {{ macro.buttonIcon('', 'pencil', 'lbl.EditMediaFolder'|trans|ucfirst, 'btn-default', { "id":"media-folder-edit", "data-toggle":"modal", "data-target":"#editMediaFolderDialog", "type":"button" } ) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig]: All id's are camelcased, so this one should be too.

File: src/Backend/Modules/Pages/Layout/Templates/Add.html.twig [hunk 4]
```diff
@@ -224,15 +224,11 @@
     <div class="col-md-12">
       <div class="btn-toolbar">
         <div class="btn-group" role="group">
-          {{ macro.cancelButton(geturl('Index')) }}
+          {{ macro.buttonIcon(geturl('Index'), 'times', 'lbl.Cancel'|trans|ucfirst) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Pages/Layout/Templates/Add.html.twig]: Add `btn-default`

File: src/Backend/Modules/Pages/Layout/Templates/Add.html.twig [hunk 5]
```diff
@@ -313,11 +309,8 @@
           </div>
         </div>
         <div class="modal-footer">
-          <button type="button" class="btn btn-default" data-dismiss="modal">
-            {{ macro.icon('times') }}{{ 'lbl.Cancel'|trans|ucfirst }}
-          </button>
-          <button id="addBlockSubmit" type="button" class="btn btn-primary">
-            {{ macro.icon('plus-square') }}{{ 'lbl.Add'|trans|ucfirst }}</button>
+          {{ macro.buttonIcon('', 'times', 'lbl.Cancel'|trans|ucfirst, 'btn-default', { "type":"button", "data-dismiss":"modal" }) }}
+          {{ macro.buttonIcon('', 'plus-square', 'lbl.Add'|trans|ucfirst, 'btn-default', { "type":"button", "id":"addBlockSubmit" }) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Pages/Layout/Templates/Add.html.twig]: Wrong class, this should be `btn-primary`

File: src/Backend/Modules/Pages/Layout/Templates/Edit.html.twig [hunk 6]
```diff
@@ -26,7 +26,7 @@
         {% if appendRevision %}
           {% set url = (url ~ '?page_revision=' ~ item.revision_id) %}
         {% endif %}
-        {{ macro.buttonIcon(url, 'eye', 'lbl.View'|trans|ucfirst, '', "target=_blank") }}
+        {{ macro.buttonIcon(url, 'eye', 'lbl.View'|trans|ucfirst, '', { "target":"_blank" }) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Pages/Layout/Templates/Edit.html.twig]: Add `btn-default`

File: src/Backend/Modules/Pages/Layout/Templates/Edit.html.twig [hunk 7]
```diff
@@ -365,21 +365,13 @@
...
         <div class="btn-group pull-left" role="group">
           {% if showPagesDelete and allowPagesDelete %}
-            <button type="button" class="btn btn-danger" data-toggle="modal" data-target="#confirmDelete">
-              {{ macro.icon('trash-o') }}
-              {{ 'lbl.Delete'|trans|ucfirst }}
-            </button>
+            {{ macro.buttonIcon('', 'trash-o', 'lbl.Delete'|trans|ucfirst, 'btn-danger', { "data-toggle":"modal", "data-target":"#confirmDelete" }) }}
           {% endif %}
-          {{ macro.cancelButton(geturl('Index')) }}
+          {{ macro.buttonIcon(geturl('Index'), 'times', 'lbl.Cancel'|trans|ucfirst) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Pages/Layout/Templates/Edit.html.twig]: Add `btn-default`

File: src/Backend/Modules/Profiles/Layout/Templates/EditGroup.html.twig [hunk 8]
```diff
@@ -23,17 +23,12 @@
...
-            </button>
+            {{ macro.buttonIcon('', 'trash-o', 'lbl.Delete'|trans|ucfirst, 'btn-danger', { "type":"button", "data-toggle":"modal", "data-target":"#confirmDelete" }) }}
           {% endif %}
           {{ macro.cancelButton(geturl('Groups')) }}
         </div>
         <div class="btn-group pull-right" role="group">
-          <button id="saveButton" type="submit" name="edit" class="btn btn-primary">
-            {{ macro.icon('floppy-o') }}{{ 'lbl.Save'|trans|ucfirst }}
-          </button>
+          {{ macro.buttonIcon('', 'floppy-o', 'lbl.Save'|trans|ucfirst, 'btn-danger', { "type":"submit", "name":"edit", "id":"saveButton" }) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Profiles/Layout/Templates/EditGroup.html.twig]: Wrong class, should be `btn-primary`

File: src/Backend/Modules/Profiles/Layout/Templates/EditProfileGroup.html.twig [hunk 9]
```diff
@@ -30,17 +30,12 @@
       <div class="btn-toolbar">
         <div class="btn-group pull-left" role="group">
           {% if showProfilesDeleteProfileGroup %}
-            <button type="button" class="btn btn-danger" data-toggle="modal" data-target="#confirmDelete">
-              {{ macro.icon('trash-o') }}
-              {{ 'lbl.Delete'|trans|ucfirst }}
-            </button>
+            {{ macro.buttonIcon('', 'trash-o', 'lbl.Delete'|trans|ucfirst, 'btn-default', { "type":"button", "data-toggle":"modal", "data-target":"#confirmDelete" }) }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Profiles/Layout/Templates/EditProfileGroup.html.twig]: Wrong lcass, should be `btn-danger`

File: src/Backend/Modules/Search/Layout/Templates/EditSynonym.html.twig [hunk 10]
```diff
@@ -65,12 +60,8 @@
...
           </div>
           <div class="modal-footer">
-            <button type="button" class="btn btn-default" data-dismiss="modal">
-              {{ macro.icon('times') }}{{ 'lbl.Cancel'|trans|ucfirst }}
-            </button>
-            <button type="submit" class="btn btn-danger">
-              {{ macro.icon('trash') }}{{ 'lbl.Delete'|trans|ucfirst }}
-            </button>
+            {{ macro.buttonIcon('', 'times', 'lbl.Cancel'|trans|ucfirst, 'btn-default', { "data-dismiss":"modal" }) }}
+            {{ macro.buttonIcon('', 'plus-trash', 'lbl.Delete'|trans|ucfirst, 'btn-danger') }}
```
Comment (mathiashelin) [inline src/Backend/Modules/Search/Layout/Templates/EditSynonym.html.twig]: Wrong icon class, should be `trash`

File: src/Backend/Modules/Blog/Layout/Templates/Add.html.twig [hunk 11]
```diff
@@ -152,15 +152,11 @@
...
+          {{ macro.buttonIcon(geturl('Index'), 'times', 'lbl.Cancel'|trans|ucfirst) }}
         </div>
         <div class="btn-group pull-right" role="group">
-          <a href="#" id="saveAsDraft" class="btn btn-default">
-            {{ macro.icon('file-o') }}{{ 'lbl.SaveDraft'|trans|ucfirst }}
-          </a>
-          <button id="addButton" type="submit" name="add" class="btn btn-primary">
-            {{ macro.icon('plus-square') }}{{ 'lbl.Add'|trans|ucfirst }}
-          </button>
+          {{ macro.buttonIcon('', 'file-o', 'lbl.SaveDraft'|trans|ucfirst, 'btn-default', {"id":"saveAsDraft"}) }}
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Blog/Layout/Templates/Add.html.twig]: Shouldn't this also contain `"type":"submit"`?

File: src/Backend/Modules/Extensions/Layout/Templates/Modules.html.twig [hunk 12]
```diff
@@ -78,12 +78,8 @@
...
           </div>
           <div class="modal-footer">
-            <button type="button" class="btn btn-default" data-dismiss="modal">
-              {{ macro.icon('times') }}{{ 'lbl.Cancel'|trans|ucfirst }}
-            </button>
-            <a href="#" class="btn btn-primary jsConfirmationSubmit">
-              {{ macro.icon('check') }}{{ 'lbl.Install'|trans|ucfirst }}
-            </a>
+            {{ macro.buttonIcon('', 'times', 'lbl.Cancel'|trans|ucfirst, 'btn-default', { "data-dismiss":"modal", "type":"button" }) }}
+            {{ macro.buttonIcon('#', 'check', 'lbl.Install'|trans|ucfirst, 'btn-primary jsConfirmationSubmit') }}
```
[the remark this answers is not in the document]
Comment (Katrienvh) [reply src/Backend/Modules/Extensions/Layout/Templates/Modules.html.twig]: The href will be filled in with js

File: src/Backend/Core/Layout/Templates/macros.html.twig [hunk 13]
```diff
@@ -1,9 +1,18 @@
...
-  </a>
+  {# When url is empty you a button, otherwise use an anchor #}
+  {% if url %}
+    <a href="{{ url }}" class="btn {{ buttonType|default('btn-default') }}" title="{{ label }}" {% for attr, value in extra %} {{ attr }}="{{ value }}"{% endfor %}>
+      {% import _self as macro %}
+      {{ macro.icon(icon) }}
+      <span class="btn-text">{{ label }}</span>
+    </a>
+  {% else %}
+    <button class="btn {{ buttonType|default('btn-default') }}" title="{{ label }}" {% for attr, value in extra %} {{ attr }}="{{ value }}"{% endfor %}>
```
[the remark this answers is not in the document]
Comment (Katrienvh) [reply src/Backend/Core/Layout/Templates/macros.html.twig]: I think it's better on one line.

File: src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemEdit.html.twig [hunk 14]
```diff
@@ -48,17 +48,12 @@
         <div class="btn-group pull-left" role="group">
           {% if showMediaLibraryMediaItemDelete %}
             {% if mediaItem.groups.count == 0 %}
-              <button type="button" class="btn btn-danger" id="blog-delete" data-toggle="modal" data-target="#confirmDelete">
-                {{ macro.icon('trash-o') }}
-                {{ 'lbl.Delete'|trans|ucfirst }}
-              </button>
+              {{ macro.buttonIcon('', 'trash-o', 'lbl.Delete'|trans|ucfirst, 'btn-danger', { "data-toggle":"modal", "type":"button", "data-target":"#confirmDelete", "id":"blog-delete" } ) }}
```
[the remark this answers is not in the document]
Comment (Katrienvh) [reply src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemEdit.html.twig]: I agree it should be fixed, but not in this PR

File: src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig (hunk 2 above)
Comment (StijnVrolijk) [reply src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig]: This id is not used at all so it should simply be removed.

File: src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig (hunk 3 above)
Comment (StijnVrolijk) [reply src/Backend/Modules/MediaLibrary/Layout/Templates/MediaItemIndex.html.twig]: This id is not used at all so it should simply be removed.

File: src/Backend/Modules/Extensions/Layout/Templates/UploadTheme.html.twig [hunk 15]
```diff
@@ -41,8 +41,7 @@
       <div class="col-md-12">
         <div class="btn-toolbar">
           <div class="btn-group pull-right" role="group">
-            <button id="importButton" type="submit" name="add" class="btn btn-primary">
-              {{ macro.icon('download') }}{{ 'lbl.Install'|trans|ucfirst }}</button>
+            {{ macro.buttonIcon('', 'download', 'lbl.Install'|trans|ucfirst, 'btn-primary', { "id":"importButton", "type":"submit", "name":"add" })
```
Comment (StijnVrolijk) [inline src/Backend/Modules/Extensions/Layout/Templates/UploadTheme.html.twig]: You're missing `}}` at the end

**Pull request merged**